lib/concurrent/atomic_reference/direct_update.rb

```diff
@@ -27,9 +27,28 @@ def update
     # @!macro [attach] atomic_reference_method_try_update
     #
     # Pass the current value to the given block, replacing it
+    # with the block's result. Return nil if the update fails.
+    #
+    # @yield [Object] Calculate a new value for the atomic reference using
+    #   given (old) value
+    # @yieldparam [Object] old_value the starting value of the atomic reference
+    #
+    # @return [Object] the new value, or nil if update failed
+    def try_update
+      old_value = get
+      new_value = yield old_value
+
+      return unless compare_and_set old_value, new_value
+
+      new_value
+    end
```

Comment:
In the end, the problem is that idiomatic Ruby abuses `nil` in ways that can make return values ambiguous. That's why I created the obligation API (#value and #reason) methods very early on. It removes the ambiguity. I don't have a strong opinion but I'm comfortable leaving it returning `nil`. For a couple of reasons. Mostly because that's the way the API previously worked. I believe that the baseline assumption is that we *do not* change APIs unless there is a compelling reason to do so. Secondly, returning `nil` is idiomatic Ruby, whether we like it or not. Finally, since the user is passing a block to the function it should be clear when looking at the code if `nil` is a valid return value, and the block can be refactored if necessary.

I'm going to merge this PR now and we can revisit later if we think there may be a better way to handle this return value.